ui/raidboss/emulator/overrides/StubbedPopupText.js

```diff
@@ -7,6 +7,22 @@ export default class StubbedPopupText extends PopupText {
 
   // Stubbed, we don't want overlay hooks
   HookOverlays() { }
+
+  // Use the current emulator timestamp unless we're using an older timestamp
```

Comment:
Sorry for the belated response, but I'm not sure I understand what's going on here.  Can you explain the different cases, and where 500 comes from?

I think this is really my only question here.

Reply:
In the `PopupText` class, `OnTriggerInternal` creates a `new Date` to pass in to `_onTriggerInternalCheckSuppressed` as well as to `_onTriggerInternalGetHelper`. As the playback in raidemulator uses the original log timestamps, that'll never line up with the `new Date` timestamp values.

https://github.com/quisquous/cactbot/blob/37be1bc0be8c67c4dea88dc449287c759bd6c968/ui/raidboss/popup-text.js#L624-L637

The override for `_onTriggerInternalCheckSuppressed` is for the check to see if a trigger is suppressed specifically, whereas the override for `_onTriggerInternalGetHelper` is for the timestamp to use for suppressing a trigger.

Without these overrides, while realtime playback from the start of the fight works fine, any seeking will cause the entire suppressed logic to fail since the timestamps passed are only a few milliseconds apart.

The ternary operator in `_onTriggerInternalCheckSuppressed` is to account for the difference between seeking (timestamp less than 500ms apart, to allow for slower PCs) vs real-time playback. It might make more sense to expose an `emulator.currentState` property that indicates `paused` vs `playing` vs `seeking` and key off of that instead.

Also, sorry for the 2-day delay in response here, I didn't get an email notification for this one.

Reply:
I think explicitly checking for state vs a heuristic based on time is strictly better, both in terms of readability but also avoiding races.

However, even with that switch, the fact that there needs to be a conditional is very suspicious to me.   Why is it not using the log timestamp at all times?

Reply:
Are you suggesting this?

```js
  _onTriggerInternalCheckSuppressed(trigger, when) {
    when = window.raidEmulator.emulator.currentEncounter.encounter.logLines[
        window.raidEmulator.emulator.currentLogLineIndex
      ].timestamp;
    return super._onTriggerInternalCheckSuppressed(trigger, when);
  }
```

There's still some potential for issues here due to the way seek batches log lines, but that potential exists with the conditional as well (basically if there's a situation where two log lines 100 lines or less apart could exceed the suppress timeout of a trigger).

Reply:
Sorry, I don't know that I have a clear suggestion here, other than saying having an if statement like that makes me both know that I really don't understand how this system is working and wondering if there's some different way that the emulator should be considering time.

Glancing at the code, it seems like part of the issue is that `OnTriggerInternal` is using the current time.  If it used a passed in time that the emulator provided, then the emulator could use a fake time throughout that it controlled.

Maybe trigger suppression needs to be fixed more thoroughly, maybe in it's own PR rather than here? Suppression is also broken during analysis, but maybe that's a totally different problem.

Reply:
> Sorry, I don't know that I have a clear suggestion here, other than saying having an if statement like that makes me both know that I really don't understand how this system is working and wondering if there's some different way that the emulator should be considering time.

The emulator always considers time according to the log line's timestamp. There's never an invocation of `new Date` that doesn't pass in a timestamp aside from this specific one for the suppress workaround.

> Glancing at the code, it seems like part of the issue is that OnTriggerInternal is using the current time. If it used a passed in time that the emulator provided, then the emulator could use a fake time throughout that it controlled.

This is the exact issue that's causing problems with the suppress code here. Realistically I'd like to modify this specific segment of code in PopupText:

https://github.com/quisquous/cactbot/blob/169e11e4a02322e3b59d718ed9e23678b352db26/ui/raidboss/popup-text.js#L624-L628

To instead read something like:

```js
  getTimestamp(trigger, matches) {
    // check matches to see if the timestamp was matched, if so return the timestamp for that
    // otherwise return `+new Date()`
  }

  OnTriggerInternal(trigger, matches) {
    const now = this.getTimestamp(trigger, matches);

    if (this._onTriggerInternalCheckSuppressed(trigger, now))
      return;
```

Then the new function could be overrode in emulator.

> Maybe trigger suppression needs to be fixed more thoroughly, maybe in it's own PR rather than here? Suppression is also broken during analysis, but maybe that's a totally different problem.

I agree that if this requires changes to non-emulator code, it should be fixed in its own PR. If that's the case, I can revert the change for suppress for now.